### PR TITLE
Enable Bzlmod for Bazel

### PR DESCRIPTION
**Pull request by meteorcloudy**
This PR enables Bzlmod as the default external dependencies system for Bazel.

Changes included:

- Enable bzlmod by default in `.bazelrc`.
- Renamed `io_bazel` to `_main` at necessary locations.
- Added lockfile `MODULE.bazel.lock`. We should enable `--lockfile_mode=error` on CI and `--lockfile_mode=update` for local builds after fixing some issues in the lock file feature, probably with 6.4.0.
- Replaced distdir with repository cache (http artifact cache) in the offline bootstrap build. This will eventually be replaced by vendoring and packaging dependencies.
- Replaced `--override_repository` hack with repository cache (http artifact cache) for running integration tests without network access. This will eventually be replaced by a true shared repository cache.
- Removed `--experimental_repository_cache_hardlinks` inside integration test to avoid `invalid cross-device link`. Because the file system of the sandbox dir is different from the repo cache dir.
- Moved common CI flags to .bazelrc for better maintenance.
- Added all Maven coords in MODULE.bazel, patched rules_jvm_external to fix some Bzlmod issues.

Next step:
- Remove WORKSPACE and clean up distdir_deps, distdir.bzl.

Fixes https://github.com/bazelbuild/bazel/issues/18957

### CODE REVIEW

File: .bazelci/postsubmit.yml [hunk 1]
```diff
@@ -1,52 +1,41 @@
...
-      - "//src:bazel"
-    test_targets:
-      - "//src/test/cpp/..."
   centos7_java11_devtoolset10:
     shell_commands:
       - sed -i.bak -e 's/^# android_sdk_repository/android_sdk_repository/' -e 's/^#
-        android_ndk_repository/android_ndk_repository/' WORKSPACE
-      - rm -f WORKSPACE.bak
+        android_ndk_repository/android_ndk_repository/' WORKSPACE.bzlmod
+      - rm -f WORKSPACE.bzlmod.bak
```
Comment (meisterT) [inline .bazelci/postsubmit.yml:8]: why are we creating this `.bak` in the first place if we delete it right away?

File: .bazelci/postsubmit.yml [hunk 2]
```diff
@@ -1,52 +1,41 @@
...
       - rm -rf $HOME/bazeltest
       - mkdir $HOME/bazeltest
     build_flags:
-      - "--test_env=TEST_INSTALL_BASE=$HOME/bazeltest/install_base"
-      - "--test_env=TEST_REPOSITORY_HOME=$OUTPUT_BASE/external"
-      - "--noremote_accept_cached"
+      - "--config=ci-linux"
```
Comment (meisterT) [inline .bazelci/postsubmit.yml:12]: Where is this config defined?

File: .bazelci/postsubmit.yml (hunk 1 above)
Comment (meteorcloudy) [reply .bazelci/postsubmit.yml:8]: It's just for compatibility with macOS, where `sed` doesn't support in-place replacing without a `.bak` file.

Comment (meteorcloudy) [reply .bazelci/postsubmit.yml:8]: I didn't know the original reason, but I guess it's just easier to keep them the same so you can search and edit all places just one time when you need some changes. We can clean this up in another PR if necessary.

**Comment by meteorcloudy** [comment]
I'm importing this PR as a CL to review, just in case it needs some tweak internally.

**Pull request closed without merging**